src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -44262,38 +44262,166 @@ static void F((object, object?) t)
 @"#nullable disable
 class C<T>
 {
 #nullable enable
     void F(object o1, object? o2, C<object> o3, C<object?> o4)
     {
-        _ /*T:object?*/ = o1;
+        _ /*T:object!*/ = o1;
         _ /*T:object?*/ = o2;
-        _ /*T:C<object!>?*/ = o3;
-        _ /*T:C<object?>?*/ = o4;
+        _ /*T:C<object!>!*/ = o3;
+        _ /*T:C<object?>!*/ = o4;
     }
 #nullable disable
     void F(C<object> o)
     {
-        _ /*T:C<object>?*/ = o;
+        _ /*T:C<object>!*/ = o;
     }
 }";
             var comp = CreateCompilation(new[] { source }, options: WithNonNullTypesTrue());
             comp.VerifyDiagnostics();
             comp.VerifyTypes();
         }
 
+        [Fact, WorkItem(36503, "https://github.com/dotnet/roslyn/issues/36503")]
+        public void Discard_Reinferred()
+        {
+            var source = @"
+#nullable enable
+interface I<in T> {}
+
+class C
+{
+    void M(I<string?> i1, I<string> i2)
+    {
+        var x = i2 ?? i1;
+        _ = x /*T:I<string!>!*/;
+
+        _ /*T:I<string!>!*/ = i2 ?? i1;
+
+        var y = (_ = i2 ?? i1);
+        _ = y /*T:I<string!>!*/;
+    }
+}";
+            var comp = CreateNullableCompilation(source);
+            comp.VerifyDiagnostics();
+            comp.VerifyTypes();
+        }
+
+        [Fact, WorkItem(36503, "https://github.com/dotnet/roslyn/issues/36503")]
+        public void Discard_Reinferred_Nested()
+        {
+            var source = @"
+#nullable enable
+interface I<in T> {}
+
+class C
+{
+    void M(I<I<string?>> i1, I<I<string>?> i2)
+    {
+        var x = i2 ?? i1;
+        _ = x /*T:I<I<string?>!>!*/;
+
+        _ /*T:I<I<string?>!>!*/ = i2 ?? i1;
+
+        var y = (_ = i2 ?? i1);
+        _ = y /*T:I<I<string?>!>!*/;
+    }
+}";
+            var comp = CreateNullableCompilation(source);
+            comp.VerifyDiagnostics();
+            comp.VerifyTypes();
+        }
+
+        [Fact, WorkItem(36503, "https://github.com/dotnet/roslyn/issues/36503")]
+        public void Discard_OutDiscard()
+        {
+            var source = @"
+class C
+{
+    void M<T>(out T t) => throw null!;
+    void M2()
+    {
+        M<string?>(out var _);
+        M<object?>(out _);
+        M<string>(out var _);
+#nullable disable annotations
+        M<object>(out _);
+    }
+}";
+            var comp = CreateNullableCompilation(source);
+            comp.VerifyDiagnostics();
+            var tree = comp.SyntaxTrees.Single();
+            var model = comp.GetSemanticModel(tree);
+            var arguments = tree.GetRoot().DescendantNodes().OfType<ArgumentSyntax>();
+
+            var discard1 = arguments.First().Expression;
+            Assert.Equal("var _", discard1.ToString());
+            Assert.Equal("System.String", model.GetTypeInfo(discard1).Type.ToTestDisplayString());
+            Assert.Equal(CodeAnalysis.NullableAnnotation.Annotated, model.GetTypeInfo(discard1).Nullability.Annotation);
+
+            var discard2 = arguments.Skip(1).First().Expression;
```

Comment:
>arguments.Skip(1).First() [](start = 27, length = 25)

Minor point: `arguments.ElementAt(1)`

Reply:
Thanks. I'll remember for next time. I'll leave as-is to meet today's deadline though.